promslog/slog.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package promslog defines standardised ways to initialize the Go standard
+// library's log/slog logger.
+// It should typically only ever be imported by main packages.
+
+//go:build go1.21
+
+package promslog
+
+import (
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+var (
+	LevelFlagOptions  = []string{"debug", "info", "warn", "error"}
+	FormatFlagOptions = []string{"logfmt", "json"}
+	StyleFlagOptions  = []string{"slog", "go-kit"}
```

Comment:
This feels strange 🤔. Why would a user decide on the logging library?

Reply:
It's not selecting the logging library, but the style of the output. The `go-kit` output still uses slog, but styles slog to look like `go-kit`.

This is meant for providing output compatibility for people who are doing, probably questionable, parsing of the log lines.

Reply:
We could also call this `promlog` style.

Reply:
Aaaah I see, but the idea of only moving to slog in 3.0 wasn't because we're allowed to make breaking changes? I'm fine with key attributes changing here 🤔

Reply:
Ah wait, we'll also change things in exporters as well right...

Reply:
Would it make sense if the library user (i.e. prometheus and exporter developers) could choose the style, but the tools user (prometheus and exporter users) could not?

I.e. make this an option in the logger constructor, but not a cmdline flag

Reply:
> make this an option in the logger constructor, but not a cmdline flag

I don't mind either way. The flag idea isn't that important to me.

Reply:
Currently, the `New()` function only returns an slog logger, no error. If we add a param for `style string`, then we should use the `AllowedStyle.Set()` method to parse/validate and set the provided style. But `AllowedStyle.Set()` can return an error -- how do we best surface that to users when creating a logger? 

Options I can think of:
- update `New()` to also return an error, if needed
- if we get an error, we just ignore and carry on setting to `slog` as default. if we do this, I can't think of a reasonably good way to surface that such a decision has been made to the caller of the function, though :thinking:

I'd vote for option 1 and am going to work towards that, unless there's strong opinions otherwise

Reply:
I'm also not sure it's overly cumbersome to have users set the style via a variable prior to logger creation to keep the current function signature for New(), eg:

```go
// default slog styling
config := &promslog.Config{}
logger := New(config)

// legacy go-kit similar styling
config := &promslog.Config{Style: &AllowedStyle{}}
if err := config.Style.Set("go-kit"); err != nil {
        // handle error
}
logger := New(config)
```

Reply:
I'd prefer the `New()` with no `error` option.

I don't actually see any use of having `Config.Style` as a settable. I don't see us needing to change style on-the-fly. In fact, as-is, the style is only adjusted at `New()` time anyway.

We could simplify this a bit.

```go
type Style string

const (
	SlogStyle Style = "slog"
	GoKitStyle Style = "go-kit"
)

...

logger := promslog.New(promslog.Config{Style: promslog.SlogStyle})
```

Reply:
> I'd prefer the New() with no error option.
> I don't actually see any use of having Config.Style as a settable. 

Same, the current design was mostly born forward from aligning it with the existing log level/format configs which are settable. I agree style doesn't need to be a settable -- it's not like we intend to maintain logging styles other than `slog` default and `go-kit`. 

> I don't see us needing to change style on-the-fly. In fact, as-is, the style is only adjusted at New() time anyway.

Correct, and I'm not even sure it's possible without writing a custom slog Handler. The style formatting is done with the `ReplaceAttr()` func in the log handler options, and when the logger is created that handler is stored unexported on the handler and I don't see any stdlib functions that can modify an existing handler

https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/log/slog/logger.go;l=112-114

At the very least, it's possible to create multiple loggers and/or recreate existing loggers, but I think that's further support it doesn't need to be settable.

Reply:
Since we all seem to appreciate the style type approach, I'll get started refactoring :+1:

Reply:
log style has been made a custom type, it's no longer settable :+1: